modules/loop_notification/loop_notification.module

```diff
@@ -5,3 +5,208 @@
+    '#title' => 'Select content types',
+    '#description' => t('Select the content types where you want notifications enabled.')
+  );
+
+  return system_settings_form($form);
+}
+
+/**
+ * Implements hook_node_insert().
+ *
+ * Invoke rules components when creating a new node.
+ */
+function loop_notification_node_insert($node) {
+  _loop_notification_notify_users($node, 'insert');
+}
+
+/**
+ * Implements hook_node_update().
+ *
+ * Invoke rules components when updating a node.
+ */
+function loop_notification_node_update($node) {
+  _loop_notification_notify_users($node, 'update');
+}
+
+/**
+ * Implements hook_comment_insert().
+ *
+ * Invoke rules components when creating a new comment.
+ */
+function loop_notification_comment_insert($comment) {
+  _loop_notification_notify_users(node_load($comment->nid), 'insert', $comment);
+}
+
+/**
+ * Implements hook_comment_update().
+ *
+ * Invoke rules components when updating a comment.
+ */
+function loop_notification_comment_update($comment) {
+  _loop_notification_notify_users(node_load($comment->nid), 'update', $comment);
+}
+
+/**
+ * Implements hook_loop_notification_message_recipient_alter().
+ *
+ * Modify the array of message recipients.
+ */
+function loop_notification_loop_notification_message_recipient_alter(&$users, $node) {
+  // Fetch users who flagged the node.
+  $query = db_select('flagging', 'f')
+    ->fields('f', array('uid'));
+  $query->innerJoin('flag', 'l', 'f.fid = l.fid', array());
+  $query->condition('f.entity_id', $node->nid,'=');
+  $query->condition('f.entity_type', 'node','=');
+
+  $result = $query->execute()->fetchCol();
+  $users = array_merge($result, $users);
+}
+
+/**
+ * Loop through recipients to send notification.
+ *
+ * @param $node
+ *   The node related to the message about to be created.
+ * @param $action
+ *   The action that triggered the notification, insert or update.
+ * @param null $comment
+ *   A comment if the action was triggered by a new or updated comment.
+ */
+function _loop_notification_notify_users($node, $action, $comment = NULL) {
+  if (_loop_notification_check_content_type($node)) {
+    // Fetch all recipients.
+    $users = _loop_notification_find_users($node);
+
+    // Only load the entity_author once.
+    $author = ($comment ? user_load($comment->uid) : user_load($node->uid));
+
+    // Notify all subscribed users.
+    foreach ($users as $user) {
+      _loop_notification_notify_user($user, $node, $action, $author, $comment);
+    }
+  }
+}
+
+/**
+ * Provide the list of users to be notified.
+ *
+ * @param $node
+ *   The node related to the message about to be created.
+ * @return array
+ *   The list of user entities that should receive the notification.
+ */
+function _loop_notification_find_users($node) {
+  $user_ids = array();
+  // Allow other modules to modify the list of recipients
+  drupal_alter('loop_notification_message_recipient', $user_ids, $node);
```

Comment:
Normally one would populate the array and then allow others to change its content.